server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java

```diff
@@ -97,68 +138,88 @@ final class BulkOperation extends ActionRunnable<BulkResponse> {
         this.relativeTimeProvider = relativeTimeProvider;
         this.indexNameExpressionResolver = indexNameExpressionResolver;
         this.client = client;
-        this.observer = new ClusterStateObserver(clusterService, bulkRequest.timeout(), logger, threadPool.getThreadContext());
+        this.observer = observer;
     }
 
     @Override
     protected void doRun() {
         assert bulkRequest != null;
         final ClusterState clusterState = observer.setAndGetObservedState();
-        if (handleBlockExceptions(clusterState)) {
+        if (handleBlockExceptions(clusterState, BulkOperation.this, this::onFailure)) {
+            return;
+        }
+        Map<ShardId, List<BulkItemRequest>> requestsByShard = groupBulkRequestsByShards(clusterState);
+        executeBulkRequestsByShard(requestsByShard, clusterState, this::redirectFailuresOrCompleteBulkOperation);
+    }
+
+    private void doRedirectFailures() {
+        assert failureStoreRedirects.isEmpty() != true : "Attempting to redirect failures, but none were present in the queue";
```

Comment:
I think that makes sense, though I think it might be better to check that during the failure handling steps as well so that we aren't queuing data that won't be needed.

Reply:
Aha, we do check it at that point already in `org.elasticsearch.action.bulk.BulkOperation#getRedirectTarget`

I'll add one more check before doing the waiting for sanity's sake though.

Reply:
I added the check to `redirectFailuresOrCompleteBulkOperation` which calls into this method.